### PR TITLE
Colonist should not try to instantiate abstract classes or interfaces

### PR DESCRIPTION
**Pull request by dkostyrev**
This PR fixes following issues in the processor:
* Colonist will no longer try to instantiate abstract classes or interfaces when settlers are produced via constructor.
* Colonist will check that `@OnProduceSettler` and `@OnAcceptSettler` (in case when `@ProduceSettlersAsClasses` is specified) callbacks have `Class<*>` parameter type
* Introduced some integration / unit tests for changed logic

### CODE REVIEW

File: colonist/processor/src/test/java/com/joom/colonist/processor/integration/ErrorReporterExtensions.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+import com.joom.colonist.processor.ErrorReporter
+import io.kotest.matchers.collections.shouldBeEmpty
+import io.kotest.matchers.collections.shouldHaveSize
+import io.kotest.matchers.shouldBe
+
+fun ErrorReporter.shouldContainErrorMessage(message: String) {
+  getErrors().shouldHaveSize(1)
+  getErrors().first().message shouldBe message
+}
```
Comment (imort) [inline colonist/processor/src/test/java/com/joom/colonist/processor/integration/ErrorReporterExtensions.kt:11]:
```suggestion
fun ErrorReporter.shouldContainSingleErrorMessage(message: String) {
  with(getErrors()) {
    shouldHaveSize(1)
    single().message shouldBe message
  }
}
```
Minor, just add more clarification and strictness

File: colonist/processor/src/test/java/com/joom/colonist/processor/integration/IntegrationTestRule.kt [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+  }
+
+  fun assertInvalidProject(sourceCodeDir: String, message: String, discoverSettlers: Boolean = true) {
+    val reporter = ErrorReporter()
+    processProject(sourceCodeDir, reporter, discoverSettlers = discoverSettlers, ignoreErrors = true)
+
+    reporter.shouldContainErrorMessage(message)
+  }
+
+  fun compileProject(sourceCodeDir: String, classpath: List<Path> = emptyList()): Path {
```
Comment (imort) [inline colonist/processor/src/test/java/com/joom/colonist/processor/integration/IntegrationTestRule.kt:64]: Minor: some functions below can be private?

**Review by imort — APPROVED** [review]
👍

File: colonist/processor/src/test/java/com/joom/colonist/processor/integration/IntegrationTestRule.kt (hunk 2 above)
Comment (dkostyrev) [reply colonist/processor/src/test/java/com/joom/colonist/processor/integration/IntegrationTestRule.kt:64]: I am planning to add tests that'll use these

**Pull request merged**